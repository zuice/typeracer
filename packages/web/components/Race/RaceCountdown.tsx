import React, { FunctionComponent, useState } from 'react';
import {
  Modal,
  ModalOverlay,
  ModalContent,
  ModalHeader,
  ModalBody,
} from '@chakra-ui/core';

import { RaceState } from '../../types/RaceState';
import { useInterval } from '../../lib/useInterval';

interface Props {
  initialCountdown: number;
  currentRaceState: RaceState;
  setCurrentRaceState: (state: RaceState) => void;
}

export const RaceCountdown: FunctionComponent<Props> = ({
  initialCountdown,
  currentRaceState,
  setCurrentRaceState,
}) => {
  const [currentCountdown, setCurrentCountdown] = useState(initialCountdown);

  useInterval(() => {
    if (currentCountdown === 0) {
      setCurrentRaceState(RaceState.IN_PROGRESS);
      clearInterval();
    } else {
      setCurrentCountdown(currentCountdown - 1);
    }
  }, 1000);

  return (
    <Modal
      isCentered
      isOpen={currentRaceState === RaceState.WAITING}
      onClose={() => {}}
    >
      <ModalOverlay />
      <ModalContent>
        <ModalHeader>You've entered the race</ModalHeader>
        <ModalBody>Race is starting... {currentCountdown}</ModalBody>
      </ModalContent>
    </Modal>
  );
};
