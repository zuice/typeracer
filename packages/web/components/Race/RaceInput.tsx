import React, {
  FunctionComponent,
  useState,
  createRef,
  useEffect,
  ChangeEvent,
  KeyboardEvent,
} from 'react';
import { Box, Input, theme } from '@chakra-ui/core';

interface Props {
  text: string;
  currentWordIndex: number;
  onSuccessfulLetter: (i: number) => void;
  onSuccessfulWord: () => void;
  onCompleted: () => void;
  disabled: boolean;
}

let currentLetterIndex = 0;
let currentGlobalIndex = 0;
let currentInputValue = '';

export const RaceInput: FunctionComponent<Props> = ({
  text,
  currentWordIndex,
  onSuccessfulLetter,
  onSuccessfulWord,
  onCompleted,
  disabled,
}) => {
  const [error, setError] = useState('');
  const [value, setValue] = useState('');
  const inputRef = createRef<HTMLInputElement>();
  const split = text.split(' ');
  const currentWord = split[currentWordIndex];
  const borderSuccess = theme.colors.green[800];
  const borderError = theme.colors.red[800];
  const hasError = error !== '';
  const borderColor = hasError ? borderError : borderSuccess;

  const setCurrentLetterIndex = (value: number) => {
    currentLetterIndex = value;
  };
  const setCurrentGlobalIndex = (value: number) => {
    currentGlobalIndex = value;
  };
  const setCurrentInputValue = (value: string) => {
    setValue(value);

    currentInputValue = value;
  };

  const handleSpacebar = (e: KeyboardEvent<HTMLInputElement>) => {
    if (currentInputValue.replace(' ', '') === currentWord) {
      e.preventDefault();

      setCurrentInputValue('');
      setCurrentLetterIndex(0);

      onSuccessfulWord();
    }
  };

  const handleKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.which === 32) {
      handleSpacebar(e);
    }
  };

  const handleCheckLetterValidity = () => {
    const letter = currentInputValue.slice(
      currentInputValue.length - 1,
      currentInputValue.length,
    );

    if (
      letter === currentWord[currentLetterIndex] &&
      letter !== '' &&
      currentWord[currentLetterIndex] !== ''
    ) {
      setCurrentLetterIndex(currentLetterIndex + 1);
      setCurrentGlobalIndex(currentGlobalIndex + 1);

      onSuccessfulLetter(currentGlobalIndex);
    }
  };

  const handleCheckBackspace = (newValue: string) => {
    const currentWordNewValueMatch = currentWord.slice(0, newValue.length);

    console.log(currentGlobalIndex);

    if (
      (newValue.length < currentInputValue.length &&
        newValue === currentWordNewValueMatch) ||
      (newValue !== '' && currentWordNewValueMatch !== '')
    ) {
      const difference = currentInputValue.length - newValue.length;

      setCurrentLetterIndex(currentLetterIndex - difference);
      setCurrentGlobalIndex(currentGlobalIndex - difference);

      onSuccessfulLetter(currentGlobalIndex);
    }
  };

  const handleCheckError = () => {
    const isInputErroneous =
      currentWord.slice(0, currentLetterIndex) !== currentInputValue;

    if (
      isInputErroneous &&
      currentWord.slice(0, currentLetterIndex) !== '' &&
      currentInputValue !== ''
    ) {
      setError('Spelling mistake!');
    } else {
      setError('');
    }
  };

  const handleCompleted = () => {
    onCompleted();

    setCurrentLetterIndex(0);
    setCurrentGlobalIndex(0);
    setCurrentInputValue('');
  };

  const handleCheckFinalWord = () => {
    const lastWordIndex = text.split(' ').length - 1;

    if (currentWordIndex === lastWordIndex) {
      if (currentInputValue === currentWord) {
        handleCompleted();
      }
    }
  };

  const handleChange = (inputValue: string) => {
    handleCheckBackspace(inputValue);
    setCurrentInputValue(inputValue);

    handleCheckLetterValidity();
    handleCheckError();
    handleCheckFinalWord();
  };

  useEffect(() => {
    inputRef.current.focus();
  });

  return (
    <>
      <Box display="flex" flex="1">
        <Box width="100vw" paddingRight="0.1rem">
          <Input
            isFullWidth
            isDisabled={disabled}
            ref={inputRef}
            name="race"
            placeholder={currentWord}
            value={value}
            borderColor={borderColor}
            onKeyDown={(e: KeyboardEvent<HTMLInputElement>) => handleKeyDown(e)}
            onChange={(e: ChangeEvent<HTMLInputElement>) =>
              handleChange(e.target.value)
            }
          />
        </Box>
      </Box>
      <Box>
        {hasError && (
          <Box as="span" color={theme.colors.red[300]}>
            Spelling mistake!
          </Box>
        )}
      </Box>
    </>
  );
};
