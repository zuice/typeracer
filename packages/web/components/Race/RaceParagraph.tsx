import { FunctionComponent } from 'react';
import { Box, theme } from '@chakra-ui/core';

interface Props {
  text: string;
  currentIndex: number;
}

export const RaceParagraph: FunctionComponent<Props> = ({
  text,
  currentIndex,
}) => {
  const split = text.split(' ');

  return (
    <Box
      padding="0.5rem"
      margin="1rem 0 1rem 0"
      border={`${theme.borders['1px']} ${theme.colors.gray[700]}`}
      borderRadius={theme.radii.md}
    >
      {split.map((word, i) =>
        i === currentIndex ? (
          <Box
            key={`${word}${i}`}
            as="span"
            color={theme.colors.green[300]}
            paddingRight="0.2rem"
          >
            {word}
          </Box>
        ) : (
          <Box key={`${word}${i}`} as="span" paddingRight="0.2rem">
            {word}
          </Box>
        ),
      )}
    </Box>
  );
};
