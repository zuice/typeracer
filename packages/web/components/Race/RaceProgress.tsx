import React, { FunctionComponent } from 'react';
import { Box, Progress } from '@chakra-ui/core';

interface Props {
  players: { username: string; progress: number }[];
}

export const RaceProgress: FunctionComponent<Props> = ({ players }) => (
  <Box display="flex" flexDirection="column">
    {players.map(player => (
      <Box key={player.username} paddingTop="1rem">
        {player.username}:
        <Progress value={player.progress} />
      </Box>
    ))}
  </Box>
);
