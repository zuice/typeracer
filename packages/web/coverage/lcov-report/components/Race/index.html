
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for components/Race</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> components/Race</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">65.94% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>91/138</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">18.75% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>6/32</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">34.62% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>9/26</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">61.48% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>75/122</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="RaceCountdown.tsx"><a href="RaceCountdown.tsx.html">RaceCountdown.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="17" class="abs high">17/17</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="3" class="abs medium">2/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="15" class="abs high">15/15</td>
	</tr>

<tr>
	<td class="file low" data-value="RaceInput.tsx"><a href="RaceInput.tsx.html">RaceInput.tsx</a></td>
	<td data-value="48.15" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 48%"></div><div class="cover-empty" style="width: 52%"></div></div>
	</td>
	<td data-value="48.15" class="pct low">48.15%</td>
	<td data-value="81" class="abs low">39/81</td>
	<td data-value="7.14" class="pct low">7.14%</td>
	<td data-value="28" class="abs low">2/28</td>
	<td data-value="13.33" class="pct low">13.33%</td>
	<td data-value="15" class="abs low">2/15</td>
	<td data-value="45.45" class="pct low">45.45%</td>
	<td data-value="77" class="abs low">35/77</td>
	</tr>

<tr>
	<td class="file high" data-value="RaceParagraph.tsx"><a href="RaceParagraph.tsx.html">RaceParagraph.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	</tr>

<tr>
	<td class="file high" data-value="RaceProgress.tsx"><a href="RaceProgress.tsx.html">RaceProgress.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	</tr>

<tr>
	<td class="file high" data-value="index.tsx"><a href="index.tsx.html">index.tsx</a></td>
	<td data-value="80.77" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 80%"></div><div class="cover-empty" style="width: 20%"></div></div>
	</td>
	<td data-value="80.77" class="pct high">80.77%</td>
	<td data-value="26" class="abs high">21/26</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="4" class="abs low">1/4</td>
	<td data-value="72.22" class="pct medium">72.22%</td>
	<td data-value="18" class="abs medium">13/18</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Sun Jan 12 2020 11:25:07 GMT-0500 (Eastern Standard Time)
            </div>
        </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    