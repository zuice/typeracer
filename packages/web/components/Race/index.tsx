import React, { FunctionComponent, useState } from 'react';
import { Box } from '@chakra-ui/core';

import { RaceState } from '../../types/RaceState';
import { RaceCountdown } from './RaceCountdown';
import { RaceParagraph } from './RaceParagraph';
import { RaceProgress } from './RaceProgress';
import { RaceInput } from './RaceInput';

interface Props {
  text: string;
  players: { username: string; progress: number }[];
}

const INITIAL_COUNTDOWN = 5;

export const Race: FunctionComponent<Props> = ({ text, players }) => {
  const [currentRaceState, setCurrentRaceState] = useState(RaceState.WAITING);
  const [meCurrentProgress, setMeCurrentProgress] = useState(0);
  const [currentWordIndex, setCurrentWordIndex] = useState(0);

  return (
    <Box>
      <RaceCountdown
        initialCountdown={INITIAL_COUNTDOWN}
        currentRaceState={currentRaceState}
        setCurrentRaceState={setCurrentRaceState}
      />
      <RaceProgress
        players={[...players, { username: 'Me', progress: meCurrentProgress }]}
      />
      <RaceParagraph text={text} currentIndex={currentWordIndex} />
      <RaceInput
        text={text}
        currentWordIndex={currentWordIndex}
        onSuccessfulLetter={(i: number) => {
          const totalLength = text.split(' ').join('').length;

          setMeCurrentProgress((i / totalLength) * 100);
        }}
        onSuccessfulWord={() => setCurrentWordIndex(currentWordIndex + 1)}
        onCompleted={() => {
          setCurrentRaceState(RaceState.COMPLETED);
          setCurrentWordIndex(0);
        }}
        disabled={currentRaceState !== RaceState.IN_PROGRESS}
      />
    </Box>
  );
};
