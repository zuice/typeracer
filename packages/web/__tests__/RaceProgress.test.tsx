import React from 'react';
import { render } from '@testing-library/react';

import { RaceProgress } from '../components/Race/RaceProgress';

describe('RaceProgress component', () => {
  it('should render', () => {
    const { asFragment } = render(
      <RaceProgress
        players={[
          { username: 'me', progress: 0 },
          { username: 'Frank', progress: 0 },
        ]}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
