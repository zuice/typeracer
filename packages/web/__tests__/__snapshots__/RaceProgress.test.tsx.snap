// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RaceProgress component should render 1`] = `
<DocumentFragment>
  <div
    class="css-j7qwjs"
  >
    <div
      class="css-17bmav7"
    >
      me:
      <div
        class="css-129h599"
      >
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="0"
          class="css-1h4aii8-Progress"
          role="progressbar"
        />
      </div>
    </div>
    <div
      class="css-17bmav7"
    >
      Frank:
      <div
        class="css-129h599"
      >
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="0"
          class="css-1h4aii8-Progress"
          role="progressbar"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
